doubly_linked_list/doubly_linked_list.py

```diff
@@ -44,25 +44,102 @@ def __len__(self):
     return self.length
 
   def add_to_head(self, value):
-    pass
-
+    self.length +=1
+    new_node = ListNode(value = value)
+    new_node.next = None
+    new_node.prev = None
+    if self.head is None:
+      self.head = new_node
+      self.tail = new_node
+    else:
+      self.head.prev = new_node
+      new_node.next = self.head
+      self.head = new_node
+    
   def remove_from_head(self):
-    pass
+    if self.head == None:
+      return
+    else:
+      self.length -=1
+      if self.head != self.tail:
+        self.head = self.head.next
+        self.head.prev = None
+      else:
+        self.head = self.tail=None
 
   def add_to_tail(self, value):
-    pass
+    self.length +=1
+    new_node = ListNode(value=value)
+    new_node.next = None
+    new_node.prev = None
+    if self.tail is None:
+      self.head = new_node
+      self.tail = new_node
+    else:
+      self.tail.next = new_node
+      new_node.prev = self.tail
+      self.tail = new_node
 
+   
   def remove_from_tail(self):
-    pass
+    if self.head is None:
+      return
+    if self.head.next is None:
+      self.head = None
+      return
+    n = self.head
+    while n.next is not None:
+      n = n.next
+    n.prev.next = None
```

Comment:
same sort of issue here. You need to return the value of the node that you are removing. I used this code:

 if not self.head and not self.tail:
____return None
self.length -= 1
if self.head == self.tail:
____current_tail = self.tail
____self.head = None
____self.tail = None
____return current_tail.value
current_tail = self.tail
self.tail = self.tail.prev
self.tail.next = None
return current_tail.value